core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateInnerUnnestWithGlobalAggregation.java

```diff
@@ -0,0 +1,365 @@
+ * <p>
+ * Transforms:
+ * <pre>
+ * - CorrelatedJoin (LEFT or INNER) on true, correlation(c1, c2)
+ *      - Input (a, c1, c2)
+ *      - Aggregation
+ *           global grouping
+ *           agg <- agg1(x)
+ *           - Projection
+ *                x <- foo(y)
+ *                - Aggregation
+ *                     group by (g)
+ *                     y <- agg2(u)
+ *                     - Unnest INNER
+ *                          g <- unnest(c1)
+ *                          u <- unnest(c2)
+ *                          replicate: ()
+ * </pre>
+ * Into:
+ * <pre>
+ * - Projection (restrict outputs)
+ *      - Aggregation
+ *           group by (a, c1, c2, unique)
+ *           agg <- agg1(x) mask(mask_symbol)
+ *           - Projection
+ *                x <- foo(y)
+ *                - Aggregation
+ *                     group by (g, a, c1, c2, unique, mask_symbol)
+ *                     y <- agg2(u)
+ *                     - Projection
+ *                          mask_symbol <- ordinality IS NOT NULL
+ *                          - Unnest LEFT with ordinality
+ *                               g <- unnest(c1)
+ *                               u <- unnest(c2)
+ *                               replicate: (a, c1, c2, unique)
+ *                               - AssignUniqueId unique
+ *                                    - Input (a, c1, c2)
+ * </pre>
+ */
+public class DecorrelateInnerUnnestWithGlobalAggregation
+        implements Rule<CorrelatedJoinNode>
+{
+    private static final Pattern<CorrelatedJoinNode> PATTERN = correlatedJoin()
+            .with(nonEmpty(correlation()))
+            .with(filter().equalTo(TRUE_LITERAL))
+            .matching(node -> node.getType() == CorrelatedJoinNode.Type.INNER || node.getType() == CorrelatedJoinNode.Type.LEFT);
+
+    @Override
+    public Pattern<CorrelatedJoinNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(CorrelatedJoinNode correlatedJoinNode, Captures captures, Context context)
+    {
+        // find global aggregation in subquery
+        List<PlanNode> globalAggregations = PlanNodeSearcher.searchFrom(correlatedJoinNode.getSubquery(), context.getLookup())
+                .where(DecorrelateInnerUnnestWithGlobalAggregation::isGlobalAggregation)
+                .recurseOnlyWhen(node -> node instanceof ProjectNode || isGlobalAggregation(node))
+                .findAll();
+
+        if (globalAggregations.isEmpty()) {
+            return Result.empty();
+        }
+
+        // if there are multiple global aggregations, the one that is closest to the source is the "reducing" aggregation, because it reduces multiple input rows to single output row
+        AggregationNode reducingAggregation = (AggregationNode) globalAggregations.get(globalAggregations.size() - 1);
+
+        // find unnest in subquery
+        Optional<UnnestNode> subqueryUnnest = PlanNodeSearcher.searchFrom(reducingAggregation.getSource(), context.getLookup())
```

Comment:
The search condition is written such that a single UNNEST is supported. Probably it could be extended to support multi-level (nested) unnesting. That could be a follow-up if we find we really need it. 
I haven't thought about it in depth, but it seems that we'd have to capture ordinality symbol from the outermost (last) UNNEST and use it analogically to how ordinality of the only UNNEST is used in this rule.